// tslint-consistent-codestyle/rules/extCurlyRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var tsutils_1 = require("tsutils");
var utils_1 = require("../src/utils");
var FAIL_MESSAGE_MISSING = "statement must be braced";
var FAIL_MESSAGE_UNNECESSARY = "unnecessary curly braces";
var OPTION_ELSE = 'else';
var OPTION_CONSISTENT = 'consistent';
var OPTION_BRACED_CHILD = 'braced-child';
var OPTION_NESTED_IF_ELSE = 'nested-if-else';
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithWalker(new ExtCurlyWalker(sourceFile, this.ruleName, {
            else: this.ruleArguments.indexOf(OPTION_ELSE) !== -1,
            consistent: this.ruleArguments.indexOf(OPTION_CONSISTENT) !== -1,
            child: this.ruleArguments.indexOf(OPTION_BRACED_CHILD) !== -1,
            nestedIfElse: this.ruleArguments.indexOf(OPTION_NESTED_IF_ELSE) !== -1,
        }));
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var ExtCurlyWalker = (function (_super) {
    tslib_1.__extends(ExtCurlyWalker, _super);
    function ExtCurlyWalker() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ExtCurlyWalker.prototype.walk = function (sourceFile) {
        var _this = this;
        var cb = function (node) {
            if (tsutils_1.isIterationStatement(node)) {
                _this._checkLoop(node);
            }
            else if (tsutils_1.isIfStatement(node)) {
                _this._checkIfStatement(node);
            }
            return ts.forEachChild(node, cb);
        };
        return ts.forEachChild(sourceFile, cb);
    };
    ExtCurlyWalker.prototype._checkLoop = function (node) {
        if (this._needsBraces(node.statement)) {
            if (node.statement.kind !== ts.SyntaxKind.Block)
                this.addFailureAtNode(node.statement, FAIL_MESSAGE_MISSING);
        }
        else if (node.statement.kind === ts.SyntaxKind.Block) {
            this._reportUnnecessary(node.statement);
        }
    };
    ExtCurlyWalker.prototype._checkIfStatement = function (node) {
        var _a = this._ifStatementNeedsBraces(node), then = _a[0], otherwise = _a[1];
        if (then) {
            if (node.thenStatement.kind !== ts.SyntaxKind.Block)
                this.addFailureAtNode(node.thenStatement, FAIL_MESSAGE_MISSING);
        }
        else if (node.thenStatement.kind === ts.SyntaxKind.Block) {
            this._reportUnnecessary(node.thenStatement);
        }
        if (otherwise) {
            if (node.elseStatement !== undefined &&
                node.elseStatement.kind !== ts.SyntaxKind.Block && node.elseStatement.kind !== ts.SyntaxKind.IfStatement)
                this.addFailureAtNode(node.elseStatement, FAIL_MESSAGE_MISSING);
        }
        else if (node.elseStatement !== undefined && node.elseStatement.kind === ts.SyntaxKind.Block) {
            this._reportUnnecessary(node.elseStatement);
        }
    };
    ExtCurlyWalker.prototype._needsBraces = function (node, allowIfElse) {
        if (tsutils_1.isBlock(node))
            return node.statements.length !== 1 || this._needsBraces(node.statements[0], allowIfElse);
        if (!allowIfElse && this.options.nestedIfElse && tsutils_1.isIfStatement(node) && node.elseStatement !== undefined)
            return true;
        if (!this.options.child)
            return false;
        if (tsutils_1.isIfStatement(node)) {
            var result = this._ifStatementNeedsBraces(node);
            return result[0] || result[1];
        }
        if (tsutils_1.isIterationStatement(node) || tsutils_1.isLabeledStatement(node))
            return this._needsBraces(node.statement);
        return node.kind === ts.SyntaxKind.SwitchStatement || node.kind === ts.SyntaxKind.TryStatement;
    };
    ExtCurlyWalker.prototype._ifStatementNeedsBraces = function (node, excludeElse) {
        if (this.options.else) {
            if (node.elseStatement !== undefined || utils_1.isElseIf(node))
                return [true, true];
        }
        else if (this.options.consistent) {
            if (this._needsBraces(node.thenStatement) ||
                !excludeElse && node.elseStatement !== undefined &&
                    (tsutils_1.isIfStatement(node.elseStatement)
                        ? this._ifStatementNeedsBraces(node.elseStatement)[0]
                        : this._needsBraces(node.elseStatement, true)))
                return [true, true];
            if (utils_1.isElseIf(node) && this._ifStatementNeedsBraces(node.parent, true)[0])
                return [true, true];
        }
        if (node.elseStatement !== undefined) {
            var statement = unwrapBlock(node.thenStatement);
            return [
                tsutils_1.isIfStatement(statement) && statement.elseStatement === undefined || this._needsBraces(statement),
                !excludeElse && this._needsBraces(node.elseStatement, true),
            ];
        }
        return [this._needsBraces(node.thenStatement), false];
    };
    ExtCurlyWalker.prototype._reportUnnecessary = function (block) {
        var closeBrace = block.getChildAt(2, this.sourceFile);
        var nextTokenStart = tsutils_1.getNextToken(closeBrace, this.sourceFile).getStart(this.sourceFile);
        var closeFix = tsutils_1.isSameLine(this.sourceFile, closeBrace.end, nextTokenStart)
            ? Lint.Replacement.deleteFromTo(closeBrace.end - 1, nextTokenStart)
            : Lint.Replacement.deleteFromTo(block.statements.end, block.end);
        this.addFailure(block.statements.pos - 1, block.end, FAIL_MESSAGE_UNNECESSARY, [
            Lint.Replacement.deleteFromTo(block.pos, block.statements.pos),
            closeFix,
        ]);
    };
    return ExtCurlyWalker;
}(Lint.AbstractWalker));
function unwrapBlock(node) {
    while (tsutils_1.isBlock(node) && node.statements.length === 1)
        node = node.statements[0];
    return node;
}
