// tslint-consistent-codestyle/rules/noElseAfterReturnRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Lint = require("tslint");
var utils_1 = require("../src/utils");
var walker_1 = require("../src/walker");
var tsutils_1 = require("tsutils");
var FAIL_MESSAGE = "unnecessary else after return";
var OPTION_ALLOW_ELSE_IF = 'allow-else-if';
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithWalker(new IfWalker(sourceFile, this.ruleName, {
            allowElseIf: this.ruleArguments.indexOf(OPTION_ALLOW_ELSE_IF) !== -1,
        }));
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var IfWalker = (function (_super) {
    tslib_1.__extends(IfWalker, _super);
    function IfWalker() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    IfWalker.prototype._checkIfStatement = function (node) {
        if (shouldCheckNode(node, this.options.allowElseIf) && endsWithReturnStatement(node.thenStatement))
            this._reportUnnecessaryElse(node.elseStatement, FAIL_MESSAGE);
    };
    return IfWalker;
}(walker_1.AbstractIfStatementWalker));
function shouldCheckNode(node, allowElseIf) {
    if (node.elseStatement === undefined)
        return false;
    if (!allowElseIf)
        return !utils_1.isElseIf(node);
    if (tsutils_1.isIfStatement(node.elseStatement) && utils_1.isElseIf(node.elseStatement))
        return false;
    while (utils_1.isElseIf(node)) {
        node = node.parent;
        if (!endsWithReturnStatement(node.thenStatement))
            return false;
    }
    return true;
}
function endsWithReturnStatement(node) {
    var end = tsutils_1.getControlFlowEnd(node);
    return end.end && end.statements.every(tsutils_1.isReturnStatement);
}
