// tslint-consistent-codestyle/rules/preferConstEnumRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var tsutils_1 = require("tsutils");
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithFunction(sourceFile, walk);
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
function walk(ctx) {
    var seen = new Set();
    var enums = [];
    var declarations = [];
    var variables = tsutils_1.collectVariableUsage(ctx.sourceFile);
    variables.forEach(function (variable, identifier) {
        if (identifier.parent.kind !== ts.SyntaxKind.EnumDeclaration || seen.has(identifier))
            return;
        var track = {
            name: identifier.text,
            isConst: tsutils_1.hasModifier(identifier.parent.modifiers, ts.SyntaxKind.ConstKeyword),
            declarations: [],
            members: new Map(),
            canBeConst: !variable.inGlobalScope && !variable.exported,
            uses: variable.uses,
        };
        for (var _i = 0, _a = variable.declarations; _i < _a.length; _i++) {
            var declaration = _a[_i];
            seen.add(declaration);
            if (declaration.parent.kind !== ts.SyntaxKind.EnumDeclaration) {
                track.canBeConst = false;
            }
            else {
                track.declarations.push(declaration.parent);
                declarations.push({
                    track: track,
                    declaration: declaration.parent
                });
            }
        }
        enums.push(track);
    });
    declarations.sort(function (a, b) { return a.declaration.pos - b.declaration.pos; });
    for (var _i = 0, declarations_1 = declarations; _i < declarations_1.length; _i++) {
        var _a = declarations_1[_i], track = _a.track, declaration = _a.declaration;
        for (var _b = 0, _c = declaration.members; _b < _c.length; _b++) {
            var member = _c[_b];
            var isConst = track.isConst ||
                member.initializer === undefined ||
                isConstInitializer(member.initializer, track.members, findEnum);
            track.members.set(tsutils_1.getPropertyName(member.name), {
                isConst: isConst,
                stringValued: isConst && member.initializer !== undefined && isStringValued(member.initializer, track.members, findEnum),
            });
            if (!isConst)
                track.canBeConst = false;
        }
    }
    for (var _d = 0, enums_1 = enums; _d < enums_1.length; _d++) {
        var track = enums_1[_d];
        if (track.isConst || !track.canBeConst || !onlyConstUses(track))
            continue;
        for (var _e = 0, _f = track.declarations; _e < _f.length; _e++) {
            var declaration = _f[_e];
            ctx.addFailure(declaration.name.pos - 4, declaration.name.end, "Enum '" + track.name + "' can be a 'const enum'.", Lint.Replacement.appendText(declaration.name.pos - 4, 'const '));
        }
    }
    function findEnum(name) {
        for (var _i = 0, enums_2 = enums; _i < enums_2.length; _i++) {
            var track = enums_2[_i];
            if (track.name !== name.text)
                continue;
            for (var _a = 0, _b = track.uses; _a < _b.length; _a++) {
                var use = _b[_a];
                if (use.location === name)
                    return track;
            }
        }
    }
}
function onlyConstUses(track) {
    for (var _i = 0, _a = track.uses; _i < _a.length; _i++) {
        var use = _a[_i];
        if (use.domain & 2 || use.domain === 1)
            continue;
        if (use.domain & 8)
            return false;
        var parent = use.location.parent;
        switch (parent.kind) {
            default:
                return false;
            case ts.SyntaxKind.ElementAccessExpression:
                if (parent.argumentExpression === undefined ||
                    parent.argumentExpression.kind !== ts.SyntaxKind.StringLiteral)
                    return false;
                break;
            case ts.SyntaxKind.PropertyAccessExpression:
        }
    }
    return true;
}
function isConstInitializer(initializer, members, findEnum) {
    return (function isConst(node, allowStrings) {
        switch (node.kind) {
            case ts.SyntaxKind.Identifier:
                var member = members.get(node.text);
                return member !== undefined && member.isConst && (allowStrings || !member.stringValued);
            case ts.SyntaxKind.StringLiteral:
                return allowStrings;
            case ts.SyntaxKind.NumericLiteral:
                return true;
            case ts.SyntaxKind.PrefixUnaryExpression:
                return isConst(node.operand, false);
            case ts.SyntaxKind.ParenthesizedExpression:
                return isConst(node.expression, allowStrings);
        }
        if (tsutils_1.isPropertyAccessExpression(node)) {
            if (!tsutils_1.isIdentifier(node.expression))
                return false;
            var track = findEnum(node.expression);
            if (track === undefined)
                return false;
            var member = track.members.get(node.name.text);
            return member !== undefined && member.isConst && (allowStrings || !member.stringValued);
        }
        if (tsutils_1.isElementAccessExpression(node)) {
            if (!tsutils_1.isIdentifier(node.expression) ||
                node.argumentExpression === undefined ||
                !tsutils_1.isStringLiteral(node.argumentExpression))
                return false;
            var track = findEnum(node.expression);
            if (track === undefined)
                return false;
            var member = track.members.get(node.argumentExpression.text);
            return member !== undefined && member.isConst && (allowStrings || !member.stringValued);
        }
        if (tsutils_1.isBinaryExpression(node))
            return node.operatorToken.kind !== ts.SyntaxKind.AsteriskAsteriskToken &&
                node.operatorToken.kind !== ts.SyntaxKind.AmpersandAmpersandToken &&
                node.operatorToken.kind !== ts.SyntaxKind.BarBarToken &&
                !tsutils_1.isAssignmentKind(node.operatorToken.kind) &&
                isConst(node.left, false) && isConst(node.right, false);
        return false;
    })(initializer, true);
}
function isStringValued(initializer, members, findEnum) {
    return (function stringValued(node) {
        switch (node.kind) {
            case ts.SyntaxKind.ParenthesizedExpression:
                return stringValued(node.expression);
            case ts.SyntaxKind.Identifier:
                return members.get(node.text).stringValued;
            case ts.SyntaxKind.PropertyAccessExpression:
                return findEnum(node.expression)
                    .members.get(node.name.text).stringValued;
            case ts.SyntaxKind.ElementAccessExpression:
                return findEnum(node.expression)
                    .members.get(node.argumentExpression.text).stringValued;
            default:
                return true;
        }
    })(initializer);
}
